### ISSUE TITLE
Please allow to skip checks with unsafe methods

### ISSUE THREAD

**Issue by glandium** [matklad/once_cell#59]
It would be useful to have unsafe methods that allow to get to the underlying values in OnceCell and Lazy, that assume initialization happened. A use case would be to avoid the initialization check when you know by other means that initialization did happen.

**Comment by matklad** [comment]
Yeah, something like this would make sense:

```
/// Safety: the cell must be in initialized state. 
unsafe fn get_unchecked(&self) -> &T
```

Could you give a specific example where this is useful though? My gut feeling is that, to use this function safely, you need to prove that initialization indeed has happened. A natural proof would be the `&T` you get from calling the initialization function. However, if you can thread a `&T` to the `get_unchecked` callsite, you don't need `get_unchecked`! And if you can't thread the `&T`, there's a high chance that the cell might not be initialized.

**Comment by matklad** [comment]
Thanks for the example! I agree that this is something which can't be expressed using only safe code, without compromising elsewhere. 

Specifically, `FooHandle` *can* be defined as `FooHandle(&'static Foo)`, and that would make this pattern fully-safe, at the expense of making `FooHandle` a non-ZST type.  If we want a ZST, we need get unchecked. BTW, another way to define `FooHandle` is `FooHandle(PhantomData<&'static Foo>)`, which maybe is better with respect to auto-traits.

My plan is to add `OnceCell::get_unchecked` but not `Lazy::get_unchecked`: `Lazy` is a purely convenience type, and I fear that adding unsafe methods to it might lead to misuse. So, the example would look like this:

```rust
pub struct Foo { ... }

#[derive(Copy, Clone)]
pub struct FooHadle(PhantomData<&'static Foo>);

static INSTANCE: sync::OnceCell<Foo> = sync::OnceCell::new();

impl Foo {
    fn new() -> Self { ... }
    fn new_handle() -> FooHandle {
        INSTANCE.get_or_init(Foo::new);
        FooHandle(PhantomData)
    }
}

impl FooHandle {
    fn get(self) -> &'static Foo {
        unsafe { INSTANCE.get_unchecked() }
    }
}

```

**Comment by glandium** [comment]
Heh, I oversimplified my actual use-case, which does, in fact, have `PhantomData<&'static Foo>`.

**Comment by xfix** [comment]
Hm, cannot `.get().unwrap_or_else(|| std::hint::unreachable_unchecked())` be used for that purpose? I think LLVM should be able to optimize that well.

**Comment by matklad** [comment]
One would think so, but apparently LLVM fails to optimize away a load: https://github.com/matklad/once_cell/pull/60#issuecomment-531489557

**Comment by matklad** [comment]
released as `1.2.0`
